controllers/pvc-mutator/encryption/keygen.go

```diff
@@ -0,0 +1,185 @@
+	// the StorageClass to make sure it's StorageOS.  The encryption label
+	// should only be added to StorageOS PVCs.
+	if !isEnabled(s.enabledLabel, pvc.GetLabels()) {
+		log.V(4).Info("pvc does not have storageos.com/encryption=true label, skipping")
+		return nil
+	}
+
+	// Do not allow secrets from another namespace to be referenced.  We assume
+	// that if the user can create PVCs in the namespace then they should be
+	// able to read volume secrets from that namespace.  We can't assume the
+	// same for other namespaces.
+	if requested, ok := pvc.Annotations[s.secretNamespaceAnnotationKey]; ok && requested != namespace {
+		return ErrCrossNamespace
+	}
+
+	// Ensure the keys exist, creating them if not.
+	nsKeyRef := s.NamespaceSecretKeyRef(namespace)
+	volKeyRef := s.VolumeSecretKeyRef(pvc, namespace)
+
+	if err := s.keys.Ensure(ctx, nsKeyRef, volKeyRef); err != nil {
+		return err
+	}
+
+	// Set annotations on the PVC pointing to the volume key secret.  The
+	// namespace key secret does not need to be passed to the control plane.
+	if pvc.Annotations == nil {
+		pvc.Annotations = make(map[string]string)
+	}
+	pvc.Annotations[s.secretNameAnnotationKey] = volKeyRef.Name
+	pvc.Annotations[s.secretNamespaceAnnotationKey] = volKeyRef.Namespace
+
+	log.Info("set volume encryption key annotations")
+	return nil
+}
+
+// NamespaceSecretKeyRef returns the reference of the secret that should be used to
+// store the user encryption key for a namespace.
+//
+// This key is used to create volume keys.
+func (s *EncryptionKeySetter) NamespaceSecretKeyRef(pvcNamespace string) client.ObjectKey {
+	return client.ObjectKey{
+		Name:      NamespaceSecretName,
+		Namespace: pvcNamespace,
+	}
+}
+
+// VolumeSecretKeyRef returns the reference of the secret that should be used to
+// store the encryption keys for a volume provisioned by the PVC.
+func (s *EncryptionKeySetter) VolumeSecretKeyRef(pvc *corev1.PersistentVolumeClaim, pvcNamespace string) client.ObjectKey {
+	annotations := pvc.GetAnnotations()
+	name, ok := annotations[s.secretNameAnnotationKey]
+	if !ok || name == "" {
+		name = VolumeSecretName()
+	}
+	namespace, ok := annotations[s.secretNamespaceAnnotationKey]
+	if !ok || namespace == "" {
+		namespace = pvcNamespace
+	}
+
+	return client.ObjectKey{
+		Name:      name,
+		Namespace: namespace,
+	}
+}
+
+// VolumeSecretName returns the name of the secret to use for the volume key.
+//
+// The secret relates to the StorageOS volume (or Kubernetes PV), not the PVC
+// which may be deleted and then the PV reused.  Since the volume hasn't been
+// provisioned yet we don't have a reference for it, so generate a unique
+// identifier to use instead.
+func VolumeSecretName() string {
+	return fmt.Sprintf("%s-%s", VolumeSecretNamePrefix, uuid.New().String())
+}
+
+// isEnabled returns true if the key is set in the kv map and its value is true.
+func isEnabled(key string, kv map[string]string) bool {
+	if kv == nil {
```

Comment:
```
Map types are reference types, like pointers or slices, and so the value of m above is nil; it doesn't point to an initialized map. A nil map behaves like an empty map when reading, but attempts to write to a nil map will cause a runtime panic
```
So `kv` is a non nil pointer, so nil check doesn't necessary (or makes any sense)

Reply:
Removed and added test